### PR TITLE
Migrate CLI from `cliff` to `argparse` 

### PR DESCRIPTION
**Pull request by cross32768**
## Motivation
To resolve problem that optuna CLI does not work for the latest importlib-metadata, an idea is to migrate CLI from `cliff` (`cliff` depends `stevedore`, which causes the above problem) to `argparse`.

## Description of the changes
Replace features depends on `cliff` by `argparse`.
~~This PR is currently WIP, but I create PR for discussion.~~

* Done
  * Rewrite all subcomands (like `optuna ask`, `optuna study optimize`, etc.) by `argparse`
  * Pass all tests in [tests/test_cli.py](https://github.com/optuna/optuna/blob/master/tests/test_cli.py)

* WIP
  * Some options implemented inside of cliff are dropped now (`--log-file`, `-h/--help`, and ` --debug`). I will implement these features next
  * `optuna help` returns help with different format. I will investigate the cause and implement format close to current style if I can

--- Original `optuna help` (depends on `cliff`) ---

```
usage: optuna [--version] [-v | -q] [--log-file LOG_FILE] [-h] [--debug]
              [--storage STORAGE]        

options:
  --version             show program's version number and exit
  -v, --verbose         Increase verbosity of output. Can be repeated.
  -q, --quiet           Suppress output except warnings and errors.
  --log-file LOG_FILE
                        Specify a file to log output. Disabled by default.
  -h, --help            Show help message and exit.
  --debug               Show tracebacks on errors.
  --storage STORAGE
                        DB URL. (e.g. sqlite:///example.db)

Commands:
  ask  Create a new trial and suggest parameters.
  best-trial  Show the best trial.
  best-trials  Show a list of trials located at the Pareto front.
  complete  print bash completion command (cliff)
  create-study  Create a new study.
  delete-study  Delete a specified study.
  help  print detailed help for another command (cliff)
  storage upgrade  Upgrade the schema of a storage.
  studies  Show a list of studies.
  study optimize  Start optimization of a study. Deprecated since version 2.0.0.
  study set-user-attr  Set a user attribute to a study.
  tell  Finish a trial, which was created by the ask command.
  trials  Show a list of trials.
```
--- Current `optuna help` (only depends on argparse) ---

```
usage: optuna [--version] {ask,best-trial,best-trials,create-study,delete-study,storage upgrade,studies,study optimize,study set-user-attr,tell,trials,help} ...

positional arguments:
  {ask,best-trial,best-trials,create-study,delete-study,storage upgrade,studies,study optimize,study set-user-attr,tell,trials,help}

options:
  --version             show program's version number and exit
```

  * For the sake of current implementation, user should enter command first, not options. 
       * For example, `optuna --storage sqlite:///example.db create-study` **does not work** on the current code (Users should use `optuna create-study--storage sqlite:///example.db`). If this seems like it's going to cause a big problem, I'll look into it.
  * Some codes have room for refactoring

** NOTE
  * Unfortunately, some features completely depends on `cliff` will be dropped (like interactive mode). I think it's unavoidable.
  * I try to keep the original behavior as much as possible, but when you find some behavior change of CLI, please tell me

### CODE REVIEW

**Comment by cross32768** [comment]
Add -h/--help and remove limitation of importlib-metadata version from setup.py

**Comment by cross32768** [comment]
--- Status Update ---
* Done
  * Rewrite all subcomands (like `optuna ask`, `optuna study optimize`, etc.) by `argparse`
  * Pass all tests in [tests/test_cli.py](https://github.com/optuna/optuna/blob/master/tests/test_cli.py)
  * Currently, command style like `optuna --storage sqlite:///example.db create-study` works well
  * Some refactoring are done
  * Update setup.py to install newer version of `importlib-metadata`

* WIP
  * Some options implemented inside of cliff are dropped now (`--log-file` and ` --debug`). I will implement these features next
  * `optuna help` returns help with different format. I will investigate the cause and implement format close to current style if I can
  * Current code does not pass CI

**Comment by cross32768** [comment]
--- Status Update (latest) ---
* Done
  * Rewrite all subcomands (like `optuna ask`, `optuna study optimize`, etc.) by `argparse`
  * Pass all tests in [tests/test_cli.py](https://github.com/optuna/optuna/blob/master/tests/test_cli.py)
  * Currently, command style like `optuna --storage sqlite:///example.db create-study` works well
  * Some refactoring are done
  * Update setup.py to install newer version of `importlib-metadata`
  * Migrate "optuna.command" from setup.py to cli.py and remove dependency to importlib-metadata

* WIP
  * Some options implemented inside of cliff are dropped now (`--log-file` and ` --debug`). I will implement these features next
  * I update by removing some needless info about cliff but still CI check does not succeed
  * Document about CLI may be deleted because it depends on cliff 
* Others
  * `optuna help` returns help with different format. An idea is to resolve it with new PR to improve help format because old document depends on cliff and we need to re-implement help format for better help.

**Comment by cross32768** [comment]
--- Status Update (latest) ---
* Done
  * Rewrite all subcomands (like `optuna ask`, `optuna study optimize`, etc.) by `argparse`
  * Pass all tests in [tests/test_cli.py](https://github.com/optuna/optuna/blob/master/tests/test_cli.py)
  * Currently, command style like `optuna --storage sqlite:///example.db create-study` works well
  * Some refactoring are done
  * Update setup.py to install newer version of `importlib-metadata`
  * Migrate "optuna.command" from setup.py to cli.py and remove dependency to importlib-metadata
  * Pass all CI check
  * `—debug`, `—log-file`, `—verbose` options is implemented

* WIP
  * Document about CLI may be deleted because it depends on cliff 
* Others
  * `optuna help` returns help with different format. An idea is to resolve it with new PR to improve help format because old document depends on cliff and we need to re-implement help format for better help.
  * Some behavior about `—debug` may be different because it depends on cliff and test_cli does not include test about `—debug`. But I think this option is not used almost all users.

**Comment by cross32768** [comment]
--- Status Update (latest) ---
* Done
  * Rewrite all subcomands (like `optuna ask`, `optuna study optimize`, etc.) by `argparse`
  * Pass all tests in [tests/test_cli.py](https://github.com/optuna/optuna/blob/master/tests/test_cli.py)
  * Currently, command style like `optuna --storage sqlite:///example.db create-study` works well
  * Some refactoring are done
  * Update setup.py to install newer version of `importlib-metadata`
  * Migrate "optuna.command" from setup.py to cli.py and remove dependency to importlib-metadata
  * Pass all CI check
  * `—debug`, `—log-file`, `—verbose` options are implemented
  * Document with sphinx-argparse is added 
    * This change format of document, but I think new document also includes all required info

* Others
  * `optuna help` returns help with different format. But recently I add --help to subcommands and it improves current help format well. Still some difference is remaining, but I think it can be ignored.

--- `optuna help` (original, depends on cliff) ---
```
usage: optuna [--version] [-v | -q] [--log-file LOG_FILE] [-h] [--debug]
              [--storage STORAGE]

options:
  --version             show program's version number and exit
  -v, --verbose         Increase verbosity of output. Can be repeated.
  -q, --quiet           Suppress output except warnings and errors.
  --log-file LOG_FILE
                        Specify a file to log output. Disabled by default.
  -h, --help            Show help message and exit.
  --debug               Show tracebacks on errors.
  --storage STORAGE
                        DB URL. (e.g. sqlite:///example.db)

Commands:
  ask  Create a new trial and suggest parameters.
  best-trial  Show the best trial.
  best-trials  Show a list of trials located at the Pareto front.
  complete  print bash completion command (cliff)
  create-study  Create a new study.
  delete-study  Delete a specified study.
  help  print detailed help for another command (cliff)
  storage upgrade  Upgrade the schema of a storage.
  studies  Show a list of studies.
  study optimize  Start optimization of a study. Deprecated since version 2.0.0.
  study set-user-attr  Set a user attribute to a study.
  tell  Finish a trial, which was created by the ask command.
  trials  Show a list of trials.
```

--- `optuna help` (this PR, depends only on argparse) ---
```
usage: optuna [-h] [--storage STORAGE] [-v | -q] [--log-file LOG_FILE] [--debug] [--version]
              {create-study,delete-study,study set-user-attr,studies,trials,best-trial,best-trials,study optimize,storage upgrade,ask,tell,help} ...

positional arguments:
  {create-study,delete-study,study set-user-attr,studies,trials,best-trial,best-trials,study optimize,storage upgrade,ask,tell,help}
    create-study        Create a new study.
    delete-study        Delete a specified study.
    study set-user-attr
                        Set a user attribute to a study.
    studies             Show a list of studies.
    trials              Show a list of trials.
    best-trial          Show the best trial.
    best-trials         Show a list of trials located at the Pareto front.
    study optimize      Start optimization of a study. Deprecated since version 2.0.0.
    storage upgrade     Upgrade the schema of a storage.
    ask                 Create a new trial and suggest parameters.
    tell                Finish a trial, which was created by the ask command.
    help                Show help message and exit.

options:
  -h, --help            show this help message and exit
  --storage STORAGE     DB URL. (e.g. sqlite:///example.db)
  -v, --verbose         Increase verbosity of output. Can be repeated.
  -q, --quiet           Suppress output except warnings and errors.
  --log-file LOG_FILE   Specify a file to log output. Disabled by default.
  --debug               Show tracebacks on errors.
  --version             show program's version number and exit
```


  * Some behavior about `—debug` may be different because it depends on cliff and test_cli does not include test about `—debug`. But I think this option is not used almost all users.

**Comment by cross32768** [comment]
At this point, the updates in this PR have done at some extent, so I'm removing the WIP from PR title (maybe some update needs more discussion).

**Comment by HideakiImamura** [comment]
@gen740 Could you review this PR?

**Review by gen740 — COMMENTED** [review]
LGTM! But, I think there is a trivial improvement.

File: optuna/cli.py
Comment (gen740) [inline optuna/cli.py:5]: I guess this `# NOQA` is no `longer` needed.
```suggestion
from argparse import ArgumentParser
from argparse import Namespace
```

**Review by gen740 — CHANGES_REQUESTED** [review]

**Comment by gonzaload** [comment]
LGTM! I also tested with my current PR #4129 and it works fine.

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:220]: Could you add a docstring for each method of `_BaseCommand`?

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:223]: Ditto.

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:946]:
```suggestion
    # Use `parent_parser` is necessary to avoid namespace conflict for -h/--help
    # between `main_parser` and `subparser`.
```

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:963]:
```suggestion
    # Some command consists of two strings.
```

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:975]:
```suggestion
            # Find the longest possible command.
```

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:1047]:
```suggestion
        # without any subcommand.
```

**Review by knshnb — COMMENTED** [review]
Thanks for the PR! I have several comments.

File: optuna/cli.py
Comment (cross32768) [reply optuna/cli.py:226]: done.

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:1018]: [Question] Where are these operations processed in the precious implementation?

File: optuna/cli.py
Comment (knshnb) [inline optuna/cli.py:969]: I wonder if we don't need to return `main_parser` because the argument is also modified. Is there any reason to do this?

Comment (cross32768) [reply optuna/cli.py]: I confirmed its redundant code. Thanking for pointing that out. Fixed.

File: optuna/cli.py
Comment (cross32768) [reply optuna/cli.py]: This code is re-implementation of `find_command` function in https://github.com/openstack/cliff/blob/master/cliff/commandmanager.py .

Comment (cross32768) [reply optuna/cli.py]: And currently I fixed code of this part for more readability.

**Comment by cross32768** [comment]
Note: I fixed the specification of `add_argument` method in `_BaseCommand` class (in the previous code, it returns `ArgumentParser` to have specification closer to `cliff`'s one, but now it returns `None`.)

**Comment by c-bata** [comment]
Let me unassign @keisuke-umezawa since this PR already got approvals from @gen740 and @gonzaload. Thank you for your reviews.

@knshnb Please merge this PR after you approved.

**Review by knshnb — APPROVED** [review]
Thanks for the update! LGTM.

**Comment by knshnb** [comment]
I wonder `_preprocess_argv` can be simpler. Let me work as follow-up work.

**Comment by knshnb** [comment]
I created a [PR](https://github.com/optuna/optuna/pull/4187).
@cross32768 Could you take a look if you have time?

**Pull request merged**
